### ISSUE TITLE
save + load model using checkpoints not working

### ISSUE THREAD

**Issue by brettkoonce** [tensorflow/swift-models#733]
running the following code:

```
import ImageClassificationModels
import TensorFlow
import Checkpoints
import Foundation

extension ResNet: Checkpointable {}

var model = ResNet(classCount: 10, depth: .resNet56, downsamplingInFirstStage: false)

let temporaryDirectory = URL(fileURLWithPath: "/tmp/cifar-save-test/")

print ("writing model to tmp directory...")
do {
  try model.writeCheckpoint(to: temporaryDirectory, name: "ResNet")
} catch {
  print(error)
}

print ("reading model from tmp directory...")
do {
  try model.readCheckpoint(from: temporaryDirectory, name: "ResNet")
} catch {
  print(error)
}

print ("successfully loaded saved model!")
```

produces the following error:

```
writing model to tmp directory...
reading model from tmp directory...
Checkpoints/Checkpointable.swift:122: Fatal error: The number of writable key paths: 330 did not match the number of tensor names: 494
Current stack trace:
0    libswiftCore.so                    0x00007f007d212760 swift_reportError + 50
1    libswiftCore.so                    0x00007f007d286d10 _swift_stdlib_reportFatalErrorInFile + 112
2    libswiftCore.so                    0x00007f007cf64f40 _assertionFailure(_:_:file:line:flags:) + 1329
3    ResNet-CIFAR10                     0x00005592dde0da9c <unavailable> + 834204
4    ResNet-CIFAR10                     0x00005592de8e8e85 <unavailable> + 12217989
5    libc.so.6                          0x00007f005a832b10 __libc_start_main + 231
6    ResNet-CIFAR10                     0x00005592dddf3a3a <unavailable> + 727610
Illegal instruction (core dumped)

```

**Comment by BradLarson** [comment]
This is due to a shortcoming in the current checkpoint reader's iteration implementation. When I first put this together, there wasn't a way to use reflection to determine if a property was writeable, so it was possible to have the reflection part of the iteration find a path that shouldn't be part of the checkpoint. As a hacky way around this, I let you specify properties that you can ignore for reading and writing via `ignoredTensorPaths` in the Checkpointable protocol.

In the case of ResNet (and anything else that uses BatchNorm), inside of BatchNorm are the properties `runningMean` and `runningVariance` that are not to be used inside a checkpoint. You need to manually ignore these by adjusting the conformance to

```swift
extension ResNet: Checkpointable {
  public var ignoredTensorPaths: Set<String> {
    return ["BatchNorm<Float>.runningMean", "BatchNorm<Float>.runningVariance"]
  }
}
```

The good news is that the new reflection enhancements upstream will allow for immutable properties to be determined, and this workaround can go away. However, because that's not supported in all of our toolchains, I held off on reworking this to use that interface.

Additionally, there's also a current flaw in the checkpoint reading where you'll need to change your code above to read

```swift
try model.readCheckpoint(from: temporaryDirectory.appendingPathComponent("ResNet"), name: "ResNet")
```

for it to find the written checkpoint. I need to align the directory naming between reading and writing checkpoints to do away with this inconsistency. Sorry about the implementation flaws here.

**Comment by brettkoonce** [comment]
@BradLarson thanks for the notes, that works here!

**Comment by brettkoonce** [comment]
working here, have trained a resnet50 model using checkpoints, thanks again for the notes!
